### ISSUE TITLE
Foundation Models

### ISSUE THREAD

**Issue by calebrob6** [torchgeo/torchgeo#1043]
### Summary

This issue is to track progress on implementing new pretrained weights from related literature into torchgeo:

- [ ] AnySat: [paper](https://arxiv.org/abs/2412.14123)
- [ ] Clay: [GitHub](https://github.com/Clay-foundation/model), [weights](https://huggingface.co/made-with-clay/Clay)
- [x] Copernicus-FM: #2646
- [x] Core-Dino: #2883
- [x] CROMA: [GitHub](https://github.com/antofuller/CROMA) #2370
- [x] DeCUR: [GitHub](https://github.com/zhu-xlab/DeCUR) #2191
- [x] Delineate Anything: #2876
- [x] DOFA: [GitHub](https://github.com/zhu-xlab/DOFA) #1903
- [x] EarthLoc: #2806
- [x] FG-MAE: [GitHub](https://github.com/zhu-xlab/FGMAE) @wangyi111, #2673
- [x] FTW: #2719
- [x] GASSL: [GitHub](https://github.com/sustainlab-group/geography-aware-ssl)
- [ ] Moonshine: [GitHub](https://github.com/moonshinelabs-ai/moonshine), [weights](https://github.com/moonshinelabs-ai/moonshine/blob/main/moonshine/models/model_parameters.py)
- [ ] OmniSat: [paper](https://arxiv.org/abs/2404.08351)
- [x] Panopticon: #2692
- [ ] Presto: [GitHub](https://github.com/nasaharvest/presto)
- [ ] Prithvi: [weights](https://huggingface.co/ibm-nasa-geospatial)
- [ ] RVSA: [GitHub](https://github.com/ViTAE-Transformer/Remote-Sensing-RVSA)
- [ ] SatMAE: [GitHub](https://github.com/sustainlab-group/SatMAE)
- [x] Satlas: [GitHub](https://github.com/allenai/satlas)
- [x] Scale-MAE: [GitHub](https://github.com/bair-climate-initiative/scale-mae) @isaaccorley, https://github.com/microsoft/torchgeo/pull/2057
- [x] SeCo: [GitHub](https://github.com/ServiceNow/seasonal-contrast)
- [ ] SkyScript: [GitHub](https://github.com/wangzhecheng/SkyScript)
- [ ] SkySense: [paper](https://arxiv.org/abs/2312.10115)
- [x] SoftCon: [paper](https://arxiv.org/abs/2405.20462v1) #2677
- [ ] SpectralGPT: [paper](https://arxiv.org/abs/2311.07113)
- [x] SSL4EO-S12: [GitHub](https://github.com/zhu-xlab/SSL4EO-S12/) #2673
- [x] SSL4EO-L: [weights](https://huggingface.co/torchgeo/ssl4eo_landsat)
- [x] Tile2Vec: #3230

and many, many more:

* https://github.com/Jack-bo1220/Awesome-Remote-Sensing-Foundation-Models
* https://github.com/xiaoaoran/awesome-RSFMs
* https://github.com/ZhanYang-nwpu/Awesome-Remote-Sensing-Multimodal-Large-Language-Model
* https://github.com/mvrl/rshf

### Rationale

[Foundation Models](https://arxiv.org/abs/2108.07258) are one of the most substantial developments in recent ML research. FMs trained on ImageNet are one of the core components of torchvision and transformers that make them so popular. TorchGeo serves as a collection of [EO FMs](https://arxiv.org/abs/2405.04285), allowing researchers to quickly and easily experiment with and design new FMs. This is critical for researchers to apply FMs to transfer learning on downstream tasks with small labeled datasets.

### Implementation

See #2057, #1903, #1884, etc. for recent PRs adding new FMs. 

If you would like to volunteer to add a particular FM, please comment on this issue to say that you're working on this.

Not sure where to get started? FMs that can be considered "multi-modal" (the same set of pre-trained weights can dynamically handle imagery from many different sensors) are the highest priority!

**Comment by nilsleh** [comment]
> * Which of the 100k or 1M models were implemented?
 
1 M models were implemented


> * This is partially done (@nilsleh do you mind commenting which you didn't import and why?)

From there we are missing `MAE` and `Date2Vec`. I remember when first trying to extract the state dict for those, I was running into some weird issue. And then I forgot about it, but I will look into it again!

**Comment by nilsleh** [comment]
> * https://github.com/ViTAE-Transformer/Remote-Sensing-RVSA

Regarding these weights, there are two things I am running into:
1. Their checkpoint contains model, optimizer states and more so at some point you would have to do `state_dict = state_dict["model"]` whereas all our other pretrained weights are just the model checkpoint already
2. Their checkpoint file was saved from GPU, and I actually don't know where I can add the `map_location=torch.device("cpu")` argument except in the torchvision source [code](https://github.com/pytorch/vision/blob/main/torchvision/models/_api.py#L92). We are calling `get_state_dict()` but it does not accept an argument for `map_location` but `load_state_dict_from_url()` does. So I think we need to open an issue with torchvision?

It feels like the above two issues could happen again down the line with other pretrained weights as well, if they don't have the license for us to upload them to huggingface in the format we prefer.

**Comment by adamjstewart** [comment]
For 1 and 2, my preference would be to modify and save only the backbone on the CPU so that users or Lightning can map it to the GPU themselves. But of course that requires a favorable license. I would start by inquiring about licenses. If there's an issue with the license we can suggest changing it and describe the use case we have in mind. If they don't respond or won't change, then and only then can we think about more complex code in TorchGeo or changes to torchvision.

**Comment by calebrob6** [comment]
Yep I think re-save the weights is the way to go.

**Comment by adamjstewart** [comment]
The list to end all lists: https://github.com/Jack-bo1220/Awesome-Remote-Sensing-Foundation-Models

**Comment by wangyi111** [comment]
I will work on adding bigger ViTs from SSL4EO-S12, FG-MAE, DeCUR, and SoftCon these days.

**Comment by Navoditamathur** [comment]
I am familiar and have worked with Prithvi, and have been experimenting with SatMAE & SatMAE++ for landcover classification

**Comment by bgheneti** [comment]
I am interested in adding support for Clay or Skyscript. I have worked with Clay before. Adding support for a CLIP-compatible model like Skyscript seems useful too.

**Comment by adamjstewart** [comment]
@bgheneti please do! Note that we have to be careful with the license. Clay is distributed under Apache-2.0, so we can't simply copy-n-paste the code. My recommendation would be to start from DOFA (already in TorchGeo) and make whatever minimal modifications are required to load the backbone weights. It doesn't matter what license the weights are distributed under, as we aren't adding them to TorchGeo, we are downloading them on the fly.

**Comment by adamjstewart** [comment]
Any time-series FMs are very high priority for us, including Prithvi, Presto, SatMAE, and SkySense

**Comment by bgheneti** [comment]
@adamjstewart curious if there are models in the repo already supporting text embedding similarity? Skyscript seems interesting from this perspective and I'm curious what you would suggest in terms of getting that sort of functionality into the repo.

**Comment by adamjstewart** [comment]
I don't believe we have any vision–language models in TorchGeo yet, but SkyScript would be the perfect one to add. The code is MIT licensed, so you could simply copy-n-paste the model code (with appropriate copyright attribution) into `torchgeo/models/skyscript.py`. See any of the recently merged PRs above (e.g., Copernicus-FM) for an example of which files to modify. I'm happy to help with adding tests.

**Comment by adamjstewart** [comment]
For anyone looking for high priority models, we would love to see more **time series** GeoFMs such as:

* AnySat
* OmniSat
* Presto
* Prithvi
* SatMAE
* SkySense
* any weather/climate FMs

**Comment by Omowonuola-Akintola** [comment]
Interested in looking into Prithvi, will explore and report back.

**Comment by adamjstewart** [comment]
@Pratik0405 and @isaaccorley also mentioned looking into Prithvi but I haven't seen any progress yet. First come first served, would love to see Prithvi support in TorchGeo! It's MIT licensed, so should be as simple as a copy-n-paste.

**Comment by Navoditamathur** [comment]
I started working on prithvi around june last year but they then released 2.0, and there are differences in mae.py and I am still looking into ways to support both

**Comment by Samuel-Vangu** [comment]
Hi,
I'd like to work on OmniSat and Tile2Vec if you don't mind.

**Comment by adamjstewart** [comment]
Tile2Vec is done, just updated the checklist.

**Comment by Samuel-Vangu** [comment]
@adamjstewart ,

I'll just work on omnisat then.

**Comment by shourya0mehta** [comment]
I'd like to work on Presto (I currently work with multi-temporal satellite data for crop yield modeling), planning to start from the DOFA integration as a reference for the loading pattern.

**Comment by shourya0mehta** [comment]
@adamjstewart before I start, Presto works on pixel time series instead of image patches, so img_size and in_chans in the weights meta don't really apply, and the transforms signature assumes images. Is there a convention for models like this, or should I just follow the existing WeightsEnum pattern and note the difference in the docs?
